dll/win32/shell32/brfolder.cpp

```diff
@@ -450,37 +458,37 @@ BrFolder_Treeview_Changed(BrFolder *info, NMTREEVIEWW *pnmtv)
 static LRESULT
 BrFolder_Treeview_Rename(BrFolder *info, NMTVDISPINFOW *pnmtv)
 {
-    WCHAR old_path[MAX_PATH], new_path[MAX_PATH];
-    NMTREEVIEWW nmtv;
-    TVITEMW item;
-
     if (!pnmtv->item.pszText)
-        return 0;
-
-    item.hItem = TreeView_GetSelection(info->hwndTreeView);
-    BrItemData *item_data = BrFolder_GetItemData(info, item.hItem);
-
-    SHGetPathFromIDListW(item_data->pidlFull, old_path);
-    lstrcpynW(new_path, old_path, _countof(new_path));
-    PathRemoveFileSpecW(new_path);
-    PathAppendW(new_path, pnmtv->item.pszText);
-
-    if (!MoveFileW(old_path, new_path))
-        return 0;
-
-    item_data->pidlFull.Free();
-    item_data->pidlChild.Free();
-    item_data->pidlFull.Attach(SHSimpleIDListFromPathW(new_path));
-    item_data->lpsfParent->ParseDisplayName(NULL, NULL, pnmtv->item.pszText, NULL,
-                                            &item_data->pidlChild, NULL);
+        return FALSE;
+    HTREEITEM hItem = TreeView_GetSelection(info->hwndTreeView);
+    BrItemData *data = BrFolder_GetItemData(info, hItem);
+    ASSERT(data);
+    ASSERT(BrFolder_GetItemAttributes(info, hItem, SFGAO_CANRENAME) & SFGAO_CANRENAME);
+
+    PITEMID_CHILD newChild;
+    HRESULT hr = data->lpsfParent->SetNameOf(info->hWnd, data->pidlChild,
+                                             pnmtv->item.pszText, SHGDN_NORMAL, &newChild);
+    if (FAILED(hr))
+        return FALSE;
 
-    item.mask = TVIF_HANDLE | TVIF_TEXT;
-    item.pszText = pnmtv->item.pszText;
-    TreeView_SetItem(info->hwndTreeView, &item);
+    LPITEMIDLIST newFull;
+    if (SUCCEEDED(hr = SHILClone((LPITEMIDLIST)data->pidlFull, &newFull)))
```

Comment:
Is this cast needed?

Reply:
Yes but I'll try to change SHILClone so we don't have to do this.